### PR TITLE
[BUG] use deep copy of bit arrays when getting array node state

### PR DESCRIPTION
**Pull request by pvditt**
Upstreaming https://github.com/unionai/flyte/pull/428
https://github.com/flyteorg/flyte/issues/5135

## Why are the changes needed?

ArrayNodeState is not supposed to get persisted unless the handle loop completes and successfully calls [PutArrayNodeState](https://github.com/unionai/flyte/blob/1fd4a0084a75797a97acf284c50dc73dea57b67f/flytepropeller/pkg/controller/nodes/array/handler.go#L661)

However, any [SetItem](https://github.com/unionai/flyte/blob/1fd4a0084a75797a97acf284c50dc73dea57b67f/flytestdlib/bitarray/compact_array.go#L39) call on a bitarray ([ex](https://github.com/unionai/flyte/blob/1fd4a0084a75797a97acf284c50dc73dea57b67f/flytepropeller/pkg/controller/nodes/array/state_store.go#L167)) that is used to store subnode phases + subtask phases will persist even if the Handle loop does not successfully call PutArrayNodeState at the end of the handle loop.

This is because the [BitSet](https://github.com/unionai/flyte/blob/1fd4a0084a75797a97acf284c50dc73dea57b67f/flytestdlib/bitarray/compact_array.go#L16) field in a bitarray is a pointer -> SetItem modifies the underlying data with an explicit call to PutArrayNodeState.

This can cause issues as we don't finish handling a subnode at a given phase if there is an error causing a return UnknownTransition after we [persist subnode state](https://github.com/unionai/flyte/blob/1fd4a0084a75797a97acf284c50dc73dea57b67f/flytepropeller/pkg/controller/nodes/array/handler.go#L366).

Example issue:

subnode "1" gets set to a terminal node phase
[eventing fails](https://github.com/unionai/flyte/blob/1fd4a0084a75797a97acf284c50dc73dea57b67f/flytepropeller/pkg/controller/nodes/array/handler.go#L644)
next handle loops has the accidentally updated state so doesn't get re-evaluated at the previous, non-terminal phase so we don't propagate the necessary eventing to set the subnode's correct phases

## How was this patch tested?
- deployed in Union 
- ran a myriad of workflows, mapping over tasks and launch plans, that had varying levels of successes, failures, and manual aborts.
- manually set eventing to fail to ensure that the next loop properly evaluated with the correct phases

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs

branched off https://github.com/flyteorg/flyte/pull/5680 instead of master to reduce conflicts while upstreaming. Will merge that in first to clear up diff.

### CODE REVIEW

File: .run/single-binary.run.xml [hunk 1]
```diff
@@ -1,13 +0,0 @@
-<component name="ProjectRunConfigurationManager">
```
[the remark this answers is not in the document]
Comment (pvditt) [reply .run/single-binary.run.xml:1]: oh thanks for catching this. Unsure how I did that. Let me add it back.

File: flytepropeller/pkg/apis/flyteworkflow/v1alpha1/zz_generated.deepcopy.go [hunk 2]
```diff
@@ -548,6 +548,10 @@ func (in *NodeStatus) DeepCopyInto(out *NodeStatus) {
 		*out = new(DynamicNodeStatus)
 		(*in).DeepCopyInto(*out)
 	}
+	if in.ArrayNodeStatus != nil {
+		in, out := &in.ArrayNodeStatus, &out.ArrayNodeStatus
+		*out = (*in).DeepCopy()
+	}
```
Comment (eapolinario) [inline flytepropeller/pkg/apis/flyteworkflow/v1alpha1/zz_generated.deepcopy.go:554]: can you document how you generated the change in this file?

File: flytepropeller/pkg/controller/config/config.go [hunk 3]
```diff
@@ -259,6 +259,7 @@ const (
 type EventConfig struct {
 	RawOutputPolicy           RawOutputPolicy `json:"raw-output-policy" pflag:",How output data should be passed along in execution events."`
 	FallbackToOutputReference bool            `json:"fallback-to-output-reference" pflag:",Whether output data should be sent by reference when it is too large to be sent inline in execution events."`
+	ErrorOnAlreadyExists      bool            `json:"error-on-already-exists" pflag:",Whether to return an error when an event already exists."`
```
Comment (eapolinario) [inline flytepropeller/pkg/controller/config/config.go:262]: worth enabling this [in single-binary](https://github.com/flyteorg/flyte/blob/ff12f36d385cdcc0ad2acc86803940cd3812ea23/charts/flyte-binary/templates/configmap.yaml#L42-L43) by default?

File: flytepropeller/pkg/apis/flyteworkflow/v1alpha1/zz_generated.deepcopy.go (hunk 2 above)
Comment (pvditt) [reply flytepropeller/pkg/apis/flyteworkflow/v1alpha1/zz_generated.deepcopy.go:554]: I did this manually on our fork when working on an array node feature a couple months ago. I couldn't get the script to run.

File: flytepropeller/pkg/controller/config/config.go (hunk 3 above)
Comment (pvditt) [reply flytepropeller/pkg/controller/config/config.go:262]: This should never really be enabled except for array node. I didn't want to update the record task event function, but I guess that might be better if users were to unknowingly change this and have executions fail.

Comment (pvditt) [reply flytepropeller/pkg/controller/config/config.go]: actually I think I can do this a little better and not expose a new config

File: flytepropeller/pkg/apis/flyteworkflow/v1alpha1/node_status.go [hunk 4]
```diff
@@ -316,6 +316,27 @@ func (in *ArrayNodeStatus) SetTaskPhaseVersion(taskPhaseVersion uint32) {
...
+		*out = *in
+	}
+}
+
+func (in *ArrayNodeStatus) DeepCopy() *ArrayNodeStatus {
+	if in == nil {
+		return nil
+	}
+
+	out := &ArrayNodeStatus{}
```
Comment (eapolinario) [inline flytepropeller/pkg/apis/flyteworkflow/v1alpha1/node_status.go:335]: Why are we copying into an empty `ArrayNodeStatus`?

File: flytepropeller/pkg/controller/config/config.go [hunk 5]
```diff
@@ -259,7 +259,8 @@ const (
 type EventConfig struct {
 	RawOutputPolicy           RawOutputPolicy `json:"raw-output-policy" pflag:",How output data should be passed along in execution events."`
 	FallbackToOutputReference bool            `json:"fallback-to-output-reference" pflag:",Whether output data should be sent by reference when it is too large to be sent inline in execution events."`
-	ErrorOnAlreadyExists      bool            `json:"error-on-already-exists" pflag:",Whether to return an error when an event already exists."`
+	// only meant to be overridden for certain node types that have different eventing behavior such as ArrayNode
+	ErrorOnAlreadyExists bool `json:"-"`
```
Comment (eapolinario) [inline flytepropeller/pkg/controller/config/config.go:263]: I understand the purpose of this now, thanks for the comment. We're adding a field to a central config object and assuming that such field will only be used in a particular situation. This feels hacky to me, but maybe there's a precedent and good reason to follow this pattern.

Comment (pvditt) [reply flytepropeller/pkg/controller/config/config.go:263]: yup - it definitely is hacky. Looking forward when/if we eventually re-work eventing in propeller

File: flytepropeller/pkg/apis/flyteworkflow/v1alpha1/node_status.go (hunk 4 above)
Comment (pvditt) [reply flytepropeller/pkg/apis/flyteworkflow/v1alpha1/node_status.go:335]: This method returns a deep copy of the ArrayNodeStatus.

This is similar to [func (in *TaskNodeStatus) DeepCopy() *TaskNodeStatus](https://github.com/flyteorg/flyte/blob/38cc76dd78e7f46283f160189cb6b38cf8fb1c8d/flytepropeller/pkg/apis/flyteworkflow/v1alpha1/node_status.go#L999)

Comment (eapolinario) [reply flytepropeller/pkg/apis/flyteworkflow/v1alpha1/node_status.go:335]: sorry, got the direction wrong. Thank for clarifying!

**Review by eapolinario — APPROVED** [review]

**Pull request merged**